### ISSUE TITLE
NumberProperty + dynamic range Property need to be set as an atomic operation

### ISSUE THREAD

**Issue by zepumph** [phetsims/axon#277]
Currently validation for setting a Property occurs immediately when setting it. This is challenging when using a `range` `Property` than can be mutated, because often the currently value could be outside of the new range, or vice versa depending on what is being set first. 

This is in part the issue explained in https://github.com/phetsims/projectile-motion/issues/192#issuecomment-554585069.

**Comment by zepumph** [comment]
The solution that @chrisklus @samreid and I came up with was to add `NumberProperty.setValueAndRange` which defers validation until both the new range and new value are set. This also helped to solve https://github.com/phetsims/projectile-motion/issues/192#issuecomment-554585069 by adding additional and similar code to NumberPropertyIO that will use `setValueAndRange` when the state of the NumberProperty is being set.

@pixelzoom can you please review the addition of `NumberProperty.setValueAndRange()`. Does this implementation seem reasonable given the constraints of wanting to change both to new values at the same time? We are also interested in your thoughts on using protected and newly marked "protected" elements from Property to achieve this.

**Comment by pixelzoom** [comment]
I can't think of a better option, and I think this is a fine approach.  But the implementation of `setValueAndRange` looks suspect.  For all of the reasons recently mentioned in https://github.com/phetsims/natural-selection/issues/43#issuecomment-572233507, you should definitely not be mutating the value of `this.range`, you should be setting it to the new value that was passed to `setValueAndRange`.  phetsims/axon#274 will also make the `this.range instanceof Property` assertion unnecessary, since `this.rangeProperty` will always exist, so recommended to do phetsims/axon#274 at the same time as this issue.   You'll also need to add unit tests to NumberPropertyTests.js.

**Comment by pixelzoom** [comment]
> We are also interested in your thoughts on using protected and newly marked "protected" elements from Property to achieve this.

I didn't reply to this question because I don't understand.

**Comment by zepumph** [comment]
I agree it makes sense to handle the above comments when working on #274. 

>I didn't reply to this question because I don't understand.

For that please look at the change in `Property.js` where we changed a private annotation to be `protected`, as well as looking more generally at all of the Property members and functions being used by NumberProperty for foul play.

**Comment by pixelzoom** [comment]
> For that please look at the change in Property.js where we changed a private annotation to be protected, ...

It looks like you're referring to `this.validate`.  Changing that to `@protected` seems appropriate.  

> ... as well as looking more generally at all of the Property members and functions being used by NumberProperty for foul play.

I don't see any evidence that you identified or addressed "foul play", but maybe I missed something.  If there's something specific you want me to review, please identify and assign back to me.

**Comment by zepumph** [comment]
Over in https://github.com/phetsims/axon/issues/274 I added unit tests. One of which demonstrated the mutation concern @pixelzoom shared in https://github.com/phetsims/axon/issues/277#issuecomment-573333881. 

@samreid, @chrisklus and I struggled a bit with this function, because the main use case is when the Property is already deferred, so we can't use that to accomplish the delay in notifications. The only solution I can see that won't just mutate the range (which I agree is less than ideal), is to exchange 
```
      this.rangeProperty.value.setMinMax( range.min, range.max );

```

with 

```
      this.rangeProperty.setPropertyValue( range );
```

which is a protected function that the NumberProperty does not have access to. I don't think we should make that public, or at least I don't want to. What do others think about a way to fix the issue with mutating the range instead of setting to the reference passed in. Note that there is a commented out unit test explaining it better.

**Comment by pixelzoom** [comment]
A few problems...

(1) Property `setPropertyValue` is documented as:

`@protected - for overriding only`

NumberProperty `setValueAndRange` is calling `this.setPropertyValue`.  That's not an override, and is therefore a violation of the API (as documented). 

(2) NumberProperty `setValueAndRange` is mutating Property values directly:

```js
super.setPropertyValue( value );
this.rangeProperty.value.setMinMax( range.min, range.max );
```

Observer functions of the form `( newValue, oldValue ) => {...}` will therefore receive an incorrect `oldValue`.  I see no documentation about that. And as noted in https://github.com/phetsims/axon/issues/277#issuecomment-573333881:
> For all of the reasons recently mentioned in phetsims/natural-selection#43 (comment), you should definitely not be mutating the value of this.range [now this.rangeProperty], you should be setting it to the new value that was passed to setValueAndRange.

If it wasn't OK to mutate Property values in Natural Selection, why is it OK to do it here, in something that is even more general?

**Comment by zepumph** [comment]
Over in https://github.com/phetsims/axon/issues/274 the phet-io team did major refactoring, adding a rangeProperty to each NumberProperty and making other optimizations. From this we were able to completely overhaul `setValueAndRange` to be based on deferring the validation to when the listeners fire. From this we no longer have either of the concerns you mentioned above. We also added tests for this. Hopefully this review will be short because the brunt of the review will be over in https://github.com/phetsims/axon/issues/274.
